includes/Core/Modules/Module.php

```diff
@@ -652,24 +653,60 @@ final protected function get_frontend_hook_output( $hook ) {
 	 * @return array List of permutations.
 	 */
 	final protected function permute_site_url( $site_url ) {
-		$urls = array();
+		$hostname = wp_parse_url( $site_url, PHP_URL_HOST );
 
-		// Get host url.
-		$host = wp_parse_url( $site_url, PHP_URL_HOST );
+		return array_reduce(
+			$this->permute_site_hosts( $hostname ),
+			function ( $urls, $host ) {
+				array_push( $urls, "https://$host", "http://$host" );
+				return $urls;
+			},
+			array()
+		);
+	}
 
-		// Add http:// and https:// to host.
-		$urls[] = 'https://' . $host;
-		$urls[] = 'http://' . $host;
+	/**
+	 * Generates common variations of the given hostname.
+	 *
+	 * Returns a list of hostnames that includes:
+	 * - (if IDN) in Punycode encoding
+	 * - (if IDN) in Unicode encoding
+	 * - with and without www. subdomain (including IDNs)
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $hostname Hostname to generate variations of.
+	 * @return string[] Hostname variations.
+	 */
+	protected function permute_site_hosts( $hostname ) {
+		$punycode = new Punycode();
+		// See \Requests_IDNAEncoder::is_ascii.
+		$is_ascii = preg_match( '/(?:[^\x00-\x7F])/', $hostname ) !== 1;
+		$is_www   = 0 === strpos( $hostname, 'www.' );
+		// Normalize hostname without www.
+		$hostname = $is_www ? substr( $hostname, strlen( 'www.' ) ) : $hostname;
+		$hosts    = array();
+		$add_host = function ( $host ) use ( &$hosts ) {
+			array_push( $hosts, $host, "www.$host" );
+		};
```

Comment:
Not a fan of inlining functions (especially in a variable) when it's not needed, as I think it makes the code harder to read. Adding items to an array is simple enough, so why not just use `array_push` directly?

If we don't want to write it multiple times, then I'd say using a foreach loop or `array_walk` or `array_reduce` would still be easier to follow.

Reply:
The clumsy syntax of older PHP makes it a bit awkward, I agree. I think we'll see this become more common in the future though with the addition of arrow functions:

```php
$hosts    = [];
$add_host = fn( $host ) => array_push( $hosts, $host, "www.$host" );
```

I like this approach because it adds them all in the same way, without the need for iteration.

I've updated it to inline this instead, which is slightly repetative but the cleanest alternative IMO.